File: Packages/NMP/Sources/NMP/Signing.swift
import Foundation
import NMPFFI

/// Immutable event body for `NMPEngine.signEvent`. NMP freezes the author
/// from the engine's active account before invoking the signer.
public struct NMPUnsignedEvent: Sendable, Hashable {
    public let createdAt: UInt64
    public let kind: UInt16
    public let tags: [[String]]
    public let content: String

    public init(
        createdAt: UInt64,
        kind: UInt16,
        tags: [[String]],
        content: String
    ) {
        self.createdAt = createdAt
        self.kind = kind
        self.tags = tags
        self.content = content
    }

    func toFfi() -> FfiSignEventRequest {
        FfiSignEventRequest(
            createdAt: createdAt,
            kind: kind,
            tags: tags,
            content: content
        )
    }
}

/// Exact verified result of a sign-only operation. Receiving this value does
/// not imply storage, publication, or creation of a write receipt.
public struct NMPSignedEvent: Sendable, Hashable {
    public let id: String
    public let pubkey: String
    public let createdAt: UInt64
    public let kind: UInt16
    public let tags: [[String]]
    public let content: String
    public let signature: String

    init(_ ffi: FfiSignedEvent) {
        id = ffi.id
        pubkey = ffi.pubkey
        createdAt = ffi.createdAt
        kind = ffi.kind
        tags = ffi.tags
        content = ffi.content
        signature = ffi.sig
    }
}

/// Translate the one-shot sign-only outcome's typed failure (#680). A signer
/// refusal is a typed `NMPError`; an engine-side `.Cancelled` becomes a Swift
/// `CancellationError` so `try await` cancellation reads naturally.
func mapSignEventFailure(_ failure: FfiSignEventFailure) -> Error {
    switch failure {
    case .SignerUnavailable(let reason): return NMPError.signerUnavailable(reason)
    case .SignerRejected(let reason): return NMPError.signerRejected(reason)
    case .InvalidSignerOutput(let reason): return NMPError.invalidSignerOutput(reason)
    case .Cancelled: return CancellationError()
    case .AlreadyConsumed: return NMPError.signEventAlreadyConsumed
    }
}

extension NMPEngine {
    /// Sign one exact event with the active signer without creating a write
    /// intent, pending row, receipt, relay plan, or publication (#680).
    ///
    /// `NmpEngine.signEvent` synchronously returns a one-shot
    /// `NmpSignEventHandle`; awaiting `handle.signed()` delivers the verified
    /// event or a typed failure. Task cancellation is wired through
    /// `withTaskCancellationHandler` to `handle.cancel()` -- MANDATORY because
    /// Swift task cancellation never reaches Rust and never interrupts the
    /// in-flight `await` (#680); the cancel wakes `signed()` to a `.Cancelled`
    /// failure, surfaced here as `CancellationError`.
    public func signEvent(_ event: NMPUnsignedEvent) async throws -> NMPSignedEvent {
        let handle = try nmpRethrowing {
            try ffi.signEvent(event: event.toFfi())
        }
        return try await withTaskCancellationHandler {
            do {
                return NMPSignedEvent(try await handle.signed())
            } catch let failure as FfiSignEventFailure {
                throw mapSignEventFailure(failure)
            } catch let error as FfiError {
                throw NMPError(error)
            }
        } onCancel: {
            handle.cancel()
        }
    }
}

File: Packages/NMPKotlin/src/main/kotlin/com/nmp/sdk/Signing.kt
package com.nmp.sdk

import kotlinx.coroutines.CancellationException
import uniffi.nmp_ffi.FfiSignEventFailure
import uniffi.nmp_ffi.FfiSignedEvent
import uniffi.nmp_ffi.FfiSignEventRequest
import uniffi.nmp_ffi.NmpEngineInterface

/** Immutable sign-only event body; NMP freezes its author from active identity. */
data class NMPUnsignedEvent(
    val createdAt: ULong,
    val kind: UShort,
    val tags: List<List<String>>,
    val content: String,
) {
    internal fun toFfi() = FfiSignEventRequest(createdAt, kind, tags, content)
}

/** Verified sign-only result; it carries no storage or publication claim. */
data class NMPSignedEvent(
    val id: String,
    val pubkey: String,
    val createdAt: ULong,
    val kind: UShort,
    val tags: List<List<String>>,
    val content: String,
    val signature: String,
) {
    internal constructor(ffi: FfiSignedEvent) : this(
        ffi.id,
        ffi.pubkey,
        ffi.createdAt,
        ffi.kind,
        ffi.tags,
        ffi.content,
        ffi.sig,
    )
}

/** Sign one exact event through the active signer (#680, #727). Starting the
 * operation either returns a one-shot
 * [uniffi.nmp_ffi.NmpSignEventHandle] or throws a synchronous start refusal;
 * after a handle exists, awaiting its `signed()` yields the fully-verified
 * event or only a typed [FfiSignEventFailure]. There is no
 * `suspendCancellableCoroutine` state
 * machine anymore: the pull IS the await. `handle.cancel()` runs in a
 * `finally` so that cancelling the calling coroutine (which drops the
 * in-flight Rust future) also withdraws the Rust operation -- Kotlin
 * coroutine cancellation never reaches Rust on its own. `cancel()` is
 * idempotent and safe after completion, so the unconditional `finally` is
 * correct on the success and failure paths too. */
internal suspend fun signEvent(
    engine: NmpEngineInterface,
    event: NMPUnsignedEvent,
): NMPSignedEvent {
    val handle = nmpRethrowing { engine.signEvent(event.toFfi()) }
    try {
        return NMPSignedEvent(handle.signed())
    } catch (failure: FfiSignEventFailure) {
        throw mapSignEventFailure(failure)
    } finally {
        handle.cancel()
    }
}

/** Translate the generated one-shot sign failure into the ergonomic surface.
 * `Cancelled` becomes a coroutine [CancellationException] (the operation was
 * withdrawn, not a signer fault); `AlreadyConsumed` is caller misuse -- a
 * second await on a one-shot handle -- surfaced as [IllegalStateException]. */
internal fun mapSignEventFailure(failure: FfiSignEventFailure): Throwable =
    when (failure) {
        is FfiSignEventFailure.SignerUnavailable ->
            NMPError.SignerUnavailable(failure.reason)
        is FfiSignEventFailure.SignerRejected ->
            NMPError.SignerRejected(failure.reason)
        is FfiSignEventFailure.InvalidSignerOutput ->
            NMPError.InvalidSignerOutput(failure.reason)
        is FfiSignEventFailure.Cancelled ->
            CancellationException("sign operation cancelled")
        is FfiSignEventFailure.AlreadyConsumed ->
            IllegalStateException("sign event handle already consumed")
    }
